File: ets2panda/compiler/base/destructuring.cpp
/**
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "destructuring.h"

#include "util/helpers.h"
#include "compiler/base/iterators.h"
#include "compiler/base/lreference.h"
#include "compiler/base/catchTable.h"
#include "compiler/core/pandagen.h"
#include "ir/base/property.h"
#include "ir/base/spreadElement.h"
#include "ir/expressions/arrayExpression.h"
#include "ir/expressions/assignmentExpression.h"
#include "ir/expressions/identifier.h"
#include "ir/expressions/objectExpression.h"

namespace panda::es2panda::compiler {
static void GenRestElement(PandaGen *pg, const ir::SpreadElement *rest_element,
                           const DestructuringIterator &dest_iterator, bool is_declaration)
{
    VReg array = pg->AllocReg();
    VReg index = pg->AllocReg();

    auto *next = pg->AllocLabel();
    auto *done = pg->AllocLabel();

    DestructuringRestIterator iterator(dest_iterator);

    // create left reference for rest element
    auto lref = JSLReference::Create(pg, rest_element, is_declaration);

    // create an empty array first
    pg->CreateEmptyArray(rest_element);
    pg->StoreAccumulator(rest_element, array);

    // index = 0
    pg->LoadAccumulatorInt(rest_element, 0);
    pg->StoreAccumulator(rest_element, index);

    pg->SetLabel(rest_element, next);

    iterator.Step(done);
    pg->StoreObjByValue(rest_element, array, index);

    // index++
    pg->LoadAccumulatorInt(rest_element, 1);
    pg->Binary(rest_element, lexer::TokenType::PUNCTUATOR_PLUS, index);
    pg->StoreAccumulator(rest_element, index);

    pg->Branch(rest_element, next);

    pg->SetLabel(rest_element, done);
    pg->LoadAccumulator(rest_element, array);

    lref.SetValue();
}

static void GenArray(PandaGen *pg, const ir::ArrayExpression *array)
{
    // RegScope rs(pg);
    DestructuringIterator iterator(pg, array);

    if (array->Elements().empty()) {
        iterator.Close(false);
        return;
    }

    TryContext try_ctx(pg);
    const auto &label_set = try_ctx.LabelSet();
    pg->SetLabel(array, label_set.TryBegin());

    for (const auto *element : array->Elements()) {
        RegScope ers(pg);

        if (element->IsRestElement()) {
            GenRestElement(pg, element->AsRestElement(), iterator, array->IsDeclaration());
            break;
        }

        // if a hole exist, just let the iterator step ahead
        if (element->IsOmittedExpression()) {
            iterator.Step();
            continue;
        }

        const ir::Expression *init = nullptr;
        const ir::Expression *target = element;

        if (element->IsAssignmentPattern()) {
            target = element->AsAssignmentPattern()->Left();
            init = element->AsAssignmentPattern()->Right();
        }

        auto lref = JSLReference::Create(pg, target, array->IsDeclaration());
        iterator.Step();

        if (init != nullptr) {
            auto *assign_value = pg->AllocLabel();
            auto *default_init = pg->AllocLabel();
            pg->BranchIfUndefined(element, default_init);
            pg->LoadAccumulator(element, iterator.Result());
            pg->Branch(element, assign_value);

            pg->SetLabel(element, default_init);
            init->Compile(pg);
            pg->SetLabel(element, assign_value);
        }

        lref.SetValue();
    }

    pg->SetLabel(array, label_set.TryEnd());

    // Normal completion
    pg->LoadAccumulator(array, iterator.Done());
    pg->BranchIfTrue(array, label_set.CatchEnd());
    iterator.Close(false);

    pg->Branch(array, label_set.CatchEnd());

    Label *end = pg->AllocLabel();
    pg->SetLabel(array, label_set.CatchBegin());
    pg->StoreAccumulator(array, iterator.Result());
    pg->LoadAccumulator(array, iterator.Done());

    pg->BranchIfTrue(array, end);
    pg->LoadAccumulator(array, iterator.Result());
    iterator.Close(true);
    pg->SetLabel(array, end);
    pg->LoadAccumulator(array, iterator.Result());
    pg->EmitThrow(array);
    pg->SetLabel(array, label_set.CatchEnd());
}

static std::tuple<const ir::Expression *, const ir::Expression *> GetAssignmentTarget(const ir::Property *prop_expr)
{
    const ir::Expression *init = nullptr;
    const ir::Expression *target = prop_expr->Value();

    if (target->IsAssignmentPattern()) {
        init = target->AsAssignmentPattern()->Right();
        target = target->AsAssignmentPattern()->Left();
    }

    return {init, target};
}

static void GenDefaultInitializer(PandaGen *pg, const ir::Expression *element, const ir::Expression *init)
{
    if (init == nullptr) {
        return;
    }

    RegScope rs(pg);
    VReg loaded_value = pg->AllocReg();
    pg->StoreAccumulator(element, loaded_value);

    auto *get_default = pg->AllocLabel();
    auto *store = pg->AllocLabel();

    pg->BranchIfUndefined(element, get_default);
    pg->LoadAccumulator(element, loaded_value);
    pg->Branch(element, store);

    // load default value
    pg->SetLabel(element, get_default);
    init->Compile(pg);

    pg->SetLabel(element, store);
}

static void GenObjectWithRest(PandaGen *pg, const ir::ObjectExpression *object, VReg rhs)
{
    const auto &properties = object->Properties();

    RegScope rs(pg);
    VReg prop_start = pg->NextReg();

    for (const auto *element : properties) {
        if (element->IsRestElement()) {
            RegScope rest_scope(pg);
            auto lref = JSLReference::Create(pg, element, object->IsDeclaration());
            pg->CreateObjectWithExcludedKeys(element, rhs, prop_start, properties.size() - 1);
            lref.SetValue();
            break;
        }

        VReg prop_reg = pg->AllocReg();

        RegScope prop_scope(pg);

        const ir::Property *prop_expr = element->AsProperty();
        const ir::Expression *key = prop_expr->Key();
        const auto [init, target] = GetAssignmentTarget(prop_expr);

        if (key->IsIdentifier()) {
            pg->LoadAccumulatorString(key, key->AsIdentifier()->Name());
        } else {
            key->Compile(pg);
        }

        pg->StoreAccumulator(key, prop_reg);

        auto lref = JSLReference::Create(pg, target, object->IsDeclaration());

        pg->LoadAccumulator(element, prop_reg);
        pg->LoadObjByValue(element, rhs);

        GenDefaultInitializer(pg, element, init);

        lref.SetValue();
    }
}

static void GenObject(PandaGen *pg, const ir::ObjectExpression *object, VReg rhs)
{
    const auto &properties = object->Properties();

    if (properties.empty() || properties.back()->IsRestElement()) {
        auto *not_nullish = pg->AllocLabel();

        pg->LoadAccumulator(object, rhs);
        pg->BranchIfCoercible(object, not_nullish);
        pg->ThrowObjectNonCoercible(object);

        pg->SetLabel(object, not_nullish);

        if (!properties.empty()) {
            return GenObjectWithRest(pg, object, rhs);
        }
    }

    for (const auto *element : properties) {
        RegScope prop_scope(pg);

        const ir::Property *prop_expr = element->AsProperty();
        const ir::Expression *key = prop_expr->Key();
        const auto [init, target] = GetAssignmentTarget(prop_expr);

        Operand prop_operand = pg->ToOwnPropertyKey(key, prop_expr->IsComputed());

        auto lref = JSLReference::Create(pg, target, object->IsDeclaration());

        if (std::holds_alternative<VReg>(prop_operand)) {
            pg->LoadAccumulator(element, std::get<VReg>(prop_operand));
            pg->LoadObjByValue(element, rhs);
        } else {
            pg->LoadAccumulator(element, rhs);
            pg->LoadObjProperty(element, prop_operand);
        }

        GenDefaultInitializer(pg, element, init);

        lref.SetValue();
    }
}

void Destructuring::Compile(PandaGen *pg, const ir::Expression *pattern)
{
    RegScope rs(pg);

    VReg rhs = pg->AllocReg();
    pg->StoreAccumulator(pattern, rhs);

    if (pattern->IsArrayPattern()) {
        GenArray(pg, pattern->AsArrayPattern());
    } else {
        GenObject(pg, pattern->AsObjectPattern(), rhs);
    }

    pg->LoadAccumulator(pattern, rhs);
}
}  // namespace panda::es2panda::compiler

File: ets2panda/linter/test/destructuring.ts
/*
 * Copyright (c) 2022-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

function destructuringParameters(): void {
  function drawText({ text = '', position: [x, y] = [0, 0] }): void {
    // Draw text
  }
  drawText({ text: 'Figure 1', position: [10, 20] });

  function print([a, b]): void {
    console.log(a, b);
  }
  print(['Hello', 'Wolrd']);

  const hello = ({
    firstName,
    lastName,
  }: {
    firstName: string;
    lastName: string;
  }): string => `Hello ${firstName} ${lastName}`;
  console.log(hello({ firstName: 'Karl', lastName: 'Marks' }));

  const person = { firstName: 'Adam', lastName: 'Smith' };
  console.log(hello(person));
}

function destructuringAssignments(): void {
  let a = 5,
    b = 10,
    c = 'value';
  ({ b, c } = { b: 200, c: 'bar' });
  [a, b] = [b, a];

  const rest: number[];
  [a, b, ...rest] = [10, 20, 30, 40, 50];

  let list = [1, 2];
  list = [...list, 3, 4];

  const e: number;
  let f: number;
  const x: { e: number };
  ({
    a,
    b: {
      c,
      d: [e, f],
    },
  } = { a: 10, b: { c: 'foo', d: [30, 40] } });
  [a, b, [x, { f }]] = [1, 2, [{ e: 20 }, { f: 5 }]];
}

function destructuringDeclarations(): void {
  const { q, w, e } = { q: 1, w: 'foo', e: true };

  function getSomeObject() {
    return { x: 1, y: 2 };
  }
  const { x, y } = getSomeObject();

  const [i, j, k] = [10, 20, 30];

  const getArray = (): number[] => [1, 2, 3];
  const [a, b] = getArray();
}

function loopVariables(): void {
  const objects: { a: number; b: string }[] = [
    { a: 10, b: 'q' },
    { a: 20, b: 'w' },
    { a: 30, b: 'e' },
  ];
  for (const { a, b } of objects) {
    console.log(a, b);
  }

  const arrays = [
    [1, 2],
    [3, 4],
    [5, 6],
  ];
  for (const [q, w] of arrays) {
    console.log(q, w);
  }

  let a: number, b: string;
  for ({ a, b } of objects) {
    console.log(a, b);
  }

  let x: number, y: number;
  for ([x, y] of arrays) {
    console.log(x, y);
  }

  const people = [
    {
      name: 'Mike Smith',
      family: { mother: 'Jane Smith', father: 'Harry Smith' },
    },
    {
      name: 'Tom Jones',
      family: { mother: 'Norah Jones', father: 'Richard Jones' },
    },
  ];
  let n: string, f: string;
  for ({
    name: n,
    family: { father: f },
  } of people) {
    console.log(`Name: ${n}, Father: ${f}`);
  }
}
